### PR TITLE
terminal: Close the terminal if EOF is detected

### LINKED ISSUE #38901: Terminal freezes in Linux session when Ctrl+C is pressed before exit
### Summary

The integrated terminal does not close properly if you press Ctrl+C before attempting to exit.

### Description

This issue does not occur when I use my mac, only while using Linux. I am using the default bash without configuration.

Steps to reproduce:
1. Launch the integrated terminal.
2. Press ^C
3. Press ^D or type `exit`
4. Observe that the terminal remains stuck in an open state (only Linux).

**Expected Behavior**: Terminal disappearing just like when you exit it.
**Actual Behavior**: Terminal remains stuck in an open state.

### Zed Version and System Specs

Zed: v0.205.5 (Zed) 
OS: Linux X11 ubuntu 22.04
Memory: 27.1 GiB
Architecture: x86_64
GPU: AMD Unknown (RADV GFX1103_R1) || radv || Mesa 23.2.1-1ubuntu3.1~22.04.3

### PR DESCRIPTION
**Pull request by lemorage**
Closes #38901

  Release Notes:

  - Fixed terminal freezing on Linux when pressing Ctrl+C followed by Ctrl+D or typing exit (#38901)

  Implementation Details:

  - Interactive shells now only close when user has sent keyboard input
  - Spawn failures (which never receive keyboard input) keep the terminal open to show error messages
  - User-initiated exits via Ctrl+D or typing "exit" close the terminal as expected
  - Tasks/commands (`Shell::WithArguments`) continue to close based on exit code (0 = success)

### CODE REVIEW

**Comment by SomeoneToIgnore** [comment]
Passing by and noting that this change is a bit odd and things were working in Linux for me before.
Could be that recent Alacritty bump could have caused this, and we need to understand what had changed and was this a regression or a new configuration is needed?

**Comment by lemorage** [comment]
Umm, possible. I never tested the before version in linux. I can grab some time see if any Alacritty related issue resulting in this.

**Comment by Veykril** [comment]
I'd like for us to see if we can make a simple test for this behavior as it feels like its easy for us to break things here (as is evident by this PR being raised now). Would you mind having a brief look at whether this is easily testable? If not I'll try to commit a bit of time into looking into that myself this week.

**Comment by lemorage** [comment]
Sure. Np!

**Comment by SomeoneToIgnore** [comment]
https://github.com/zed-industries/zed/pull/39232 adds a test that fails on Linux, to help with fixing the issue.

**Comment by lemorage** [comment]
Thanks! I'll see!

File: crates/terminal/src/terminal.rs [hunk 1]
```diff
@@ -1929,13 +1929,31 @@ impl Terminal {
...
+
+                let should_close = if error_code.is_none() {
+                    let is_oneshot_command =
+                        matches!(self.template.shell, Shell::WithArguments { .. });
+
+                    if is_oneshot_command {
+                        /* Close only if success or no exit code */
+                        self.child_exited.is_none_or(|e| e.code() == Some(0))
+                    } else {
+                        /* Always close the interactive shell */
```
Comment (Veykril) [inline crates/terminal/src/terminal.rs]: I think this will cause us to close the shell again if we fail to spawn it for whatever reason. Now, that shouldn't really be happening but an example would be a `$SHELL` being misconfigured and us using `Shell::System`, which we then again don't surface to the user because of the immediate closing of the shell

Comment (lemorage) [reply crates/terminal/src/terminal.rs]: Valid concern. But tbh, a fundamental problem is that we cannot distinguish between:
  - PTY Exit from spawn failure
  - PTY Exit from user pressing Ctrl+D
So I find it hard to have the best of both worlds. Do u have any insights on this?

Comment (Veykril) [reply crates/terminal/src/terminal.rs]: Hmm, that is problematic. Is there really no way for us to differentiate spawn failure from a EOF signal here? ...

File: crates/terminal/src/terminal.rs [hunk 2]
```diff
@@ -2096,13 +2100,30 @@ impl Terminal {
...
             None => {
-                if self.child_exited.is_none_or(|e| e.code() == Some(0)) {
+                // Always record the latest exit status.
+                if let Some(e) = e {
+                    self.child_exited = Some(e);
+                }
+
+                let should_close = if error_code.is_none() {
+                    let is_interactive_shell =
+                        !matches!(self.template.shell, Shell::WithArguments { .. });
```
Comment (Veykril) [inline crates/terminal/src/terminal.rs]: This doesn't work. Either of the three variants could be an interactive shell. A better approach i think would be to check if `self.task.is_none()`, then we are definitely a shell session and not a task

Comment (lemorage) [reply crates/terminal/src/terminal.rs]: Good idea! I'll see.

File: crates/terminal/src/terminal.rs [hunk 3]
```diff
@@ -2472,28 +2490,30 @@ mod tests {
...
-        let wakeup = event_rx.recv().await.expect("No wakeup event received");
-        assert_eq!(wakeup, first_event, "Expected wakeup, got {wakeup:?}");
-
         terminal.update(cx, |terminal, _| {
             let success = terminal.try_keystroke(&Keystroke::parse("ctrl-c").unwrap(), false);
             assert!(success, "Should have registered ctrl-c sequence");
         });
+
+        // Give shell time to process SIGINT
+        smol::Timer::after(Duration::from_millis(50)).await;
```
Comment (Veykril) [inline crates/terminal/src/terminal.rs]: This seems a bit odd, we have the event receiving logic from before expecting a wakeup after all. We should keep that instead, if the events differ on platforms we might want to make this a bit more flexible, but waiting for a given duration makes the test flaky

File: crates/terminal/src/terminal.rs [hunk 4]
```diff
@@ -2096,13 +2100,26 @@ impl Terminal {
...
         }
-        if let Some(e) = e {
-            self.child_exited = Some(e);
-        }
+
         let task = match &mut self.task {
             Some(task) => task,
             None => {
-                if self.child_exited.is_none_or(|e| e.code() == Some(0)) {
+                // Always record the latest exit status.
```
Comment (Veykril) [inline crates/terminal/src/terminal.rs:2247]: Can we add a comment to this entire block describing the intent of this PR? Otherwise this logic looks arbitrary to readers and we might regress unintentionally

File: crates/terminal/src/terminal.rs (hunk 3 above)
Comment (lemorage) [reply crates/terminal/src/terminal.rs]: I remembered this test had some issues on Linux. I'll check this later on my Linux machine.

File: crates/terminal/src/terminal.rs (hunk 4 above)
Comment (lemorage) [reply crates/terminal/src/terminal.rs:2247]: Sure thing!

**Comment by lemorage** [comment]
After Ctrl + C and Ctrl + D, the terminal exits smoothly on Linux. This is verified. But in the `test_terminal_eof`, whatever approaches I try, it always fails. It works only after a short interval here, say `smol::Timer::after(Duration::from_millis(50)).await;`. So I still keep it there in the test. Feel free to tell if you have any insights.

File: crates/terminal/src/terminal.rs [hunk 5]
```diff
@@ -2465,22 +2493,36 @@ mod tests {
...
         })
         .detach();
 
+        // First event should be Wakeup or BreadcrumbsChanged, depending on OS
+        #[cfg(target_os = "macos")]
         let first_event = Event::Wakeup;
-        let wakeup = event_rx.recv().await.expect("No wakeup event received");
-        assert_eq!(wakeup, first_event, "Expected wakeup, got {wakeup:?}");
+        #[cfg(target_os = "linux")]
+        let first_event = Event::BreadcrumbsChanged;
```
Comment (Veykril) [inline crates/terminal/src/terminal.rs]: Given this is weirdly the first event on linux, could you check if the timing issue disappears if we make it poll events until the first wakeup event comes in? I unfortunately do not have a linux machine set up at the moment so I can't test this myself.

**Review by Veykril — COMMENTED** [review]
Sorry for the delay, turns out I never submitted this review

**Comment by lemorage** [comment]
I did some time of testing. So if we poll events until the first wakeup event comes in, the test still fails on Linux unless we adds the following:
```rs
while let Ok(Ok(_first_event)) = smol_timeout(Duration::from_millis(100), event_rx.recv()).await {}
```

But this is only a remedy on Linux end, which will in turn cause tests failing on MacOS. So whatever a special case `#[cfg(target_os = "linux")]` is unavoidable, per my cents. Or I would like to hear if there is anything I am missing.

**Review by smitbarmase — CHANGES_REQUESTED** [review]
Sorry it took me a while to check this.

The changes look good to me overall, except for the `test_terminal_eof` part. Could you rebase on the latest `main` and check whether it still fails on Linux without any changes?

From my testing on Linux, it passes fine even with the iterations bumped to 100 and across multiple runs. It’s probably possible that recent scheduler changes or alacritty changes improved things a bit.

I think we can simplify that test quite a bit based on this.

FWIW: I can reproduce this with bash as shell on both macOS and Linux. I haven't tested this patch yet, but it could solve this issue on both platforms.

**Comment by smitbarmase** [comment]
Hey @lemorage, any findings after rebase?

File: crates/terminal/src/terminal.rs [hunk 6]
```diff
@@ -2240,7 +2244,32 @@ impl Terminal {
         let task = match &mut self.task {
             Some(task) => task,
             None => {
-                if self.child_exited.is_none_or(|e| e.code() == Some(0)) {
+                // Always record the latest exit status.
+                if let Some(e) = exit_status {
```
Comment (smitbarmase) [inline crates/terminal/src/terminal.rs:2248]: We already do this on line 2241 right? Why do we need this again right after it?

File: crates/terminal/src/terminal.rs [hunk 7]
```diff
@@ -2240,7 +2244,32 @@ impl Terminal {
...
+                // For interactive shells (no task), we need to differentiate between:
+                // 1. Shell spawn failures (e.g., misconfigured $SHELL) - should NOT close
+                // 2. User-initiated exits (Ctrl+D, typing "exit") - should close
+                //
+                // We use keyboard_input_sent as the differentiator: spawn failures never
+                // receive keyboard input, while user exits always require it.
+                //
+                // For tasks/commands, we close on successful exit (exit code 0).
+                let should_close = if raw_status.is_none() {
+                    if self.task.is_none() {
```
Comment (smitbarmase) [inline crates/terminal/src/terminal.rs:2263]: We already know we are in `task` is `None` case, so this is always true?

File: crates/terminal/src/terminal.rs [hunk 8]
```diff
@@ -2240,7 +2244,32 @@ impl Terminal {
...
+                //
+                // For interactive shells (no task), we need to differentiate between:
+                // 1. Shell spawn failures (e.g., misconfigured $SHELL) - should NOT close
+                // 2. User-initiated exits (Ctrl+D, typing "exit") - should close
+                //
+                // We use keyboard_input_sent as the differentiator: spawn failures never
+                // receive keyboard input, while user exits always require it.
+                //
+                // For tasks/commands, we close on successful exit (exit code 0).
+                let should_close = if raw_status.is_none() {
```
Comment (smitbarmase) [inline crates/terminal/src/terminal.rs:2262]: Also, `raw_status` is never `None` (if you check the call sites) so everything inside it is just a dead code.

**Review by smitbarmase — COMMENTED** [review]
Hey, thanks for this. Looking at the code, it seems like a bunch of cases don’t really make sense. Also, from testing, it doesn’t do what it claims to do, since the `keyboard_input_sent` heuristic is unreachable.

I’m sure this may have drifted during a rebase or changed further on top of the original patch. I’m going to close this for now, but feel free to take another shot at it. I think we should have tests covering every branch this might introduce.

After checking `test_terminal_eof` more, it is indeed flaky on Linux because we call `Ctrl-C` and `Ctrl-D` right after each other. Sometimes `Ctrl-C` causes consuming/flushing behavior, which makes the test behave inconsistently. This needs special attention to fix, and I think it’s a separate issue from this patch.

A better way to test the behavior we want would be to run `false`, and then `exit`. That also blocks the terminal from closing, and it reproduces on both macOS and Linux with bash.
